src/inmanta/server/services/paramservice.py

```diff
@@ -181,9 +184,12 @@ async def _update_param(
         else:
             param = params[0]
             value_updated = param.value != value
-            await param.update(
-                source=source, value=value, updated=datetime.datetime.now().astimezone(), metadata=metadata, expires=expires
-            )
+            if expires is not None:
```

Comment:
The if-block that belongs to this else-block contains a type violations. The constructor `data.Parameter` doesn't allow the `expires` field to be `None`. Did you run mypy on this? Do you see why we didn't catch this in a test case. Are we missing test coverage here?

Reply:
> Did you run mypy on this?
Yes, mypy seems ok with that, weirdly enough
I'm looking into this

Reply:
It seems the constructor is fine with the `expires` field being `None` e.g.:

```python
from uuid import UUID

from inmanta import data

param = data.Parameter(
    environment=UUID("0d54b1ac-6ddf-4e18-8c8f-9ac30d4d58ac"),
    name="name",
    source="source",
    # resource_id=resource_id,
    # value=value,
    # updated=datetime.datetime.now().astimezone(),
    # metadata=metadata,
    expires=None,
)

assert param.expires is None
```

Reply:
Strange. But it's still a typing violation. Now I also see that the data migration script is also missing a `NOT NULL` constraint. That needs to be fixed as well. If you would not do that in this PR, you should create a ticket for this.

Reply:
Indeed. Also I think the migration is incomplete. Parameters (not associated with resource id) should get a default value expires=False and Facts (associated with a resource id) should get a default value of True ?

Reply:
Created a ticket here: https://github.com/inmanta/inmanta-core/issues/7045

Reply:
I think this is another type violation: https://github.com/inmanta/inmanta-core/blob/636f3f12569fb2be6296c831eb3318a00bbe5b02/src/inmanta/server/services/paramservice.py#L250
resource_id is None for facts and _update_param should allow None for this parameter.

I'll try to clean this up by adding a proper set_fact endpoint

Reply:
Ticket to add the endpoints here: https://github.com/inmanta/inmanta-core/issues/7068